zulipterminal/ui_tools/boxes.py

```diff
@@ -807,12 +815,23 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
                         "Cannot narrow to message without specifying recipients."
                     )
         elif is_command_key("GO_BACK", key):
+            saved_draft = self.model.session_draft_message()
             self.send_stop_typing_status()
-            self._set_compose_attributes_to_defaults()
-            self.view.controller.exit_editor_mode()
-            self.main_view(False)
-            self.view.middle_column.set_focus("body")
-            self.exit_compose_box()
+
+            compose_not_in_edit_mode = self.msg_edit_state is None
+            compose_message_length = len(self.msg_write_box.edit_text)
+            exceeds_max_length = (
+                compose_message_length >= MAX_MESSAGE_LENGTH_CONFIRMATION_POPUP
+            )
+            not_saved_as_draft = (
+                saved_draft is None
+                or self.msg_write_box.edit_text != saved_draft.get("content")
+            )
+
+            if compose_not_in_edit_mode and exceeds_max_length and not_saved_as_draft:
+                self.view.controller.exit_compose_confirmation_popup()
+            else:
+                self.exit_compose_box()
```

Comment:
This logic is a lot clearer :+1: I hope you agree :)

Nit: consider a variable for `...edit_text`, perhaps instead of the message length?

Nit: group the initial/middle data setup (saved_draft, compose_max_length/equivalent above), then the derived bools - or each prep+bool together, if not reused.